File: nextGreaterEle.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No

// we'll add elements index in stack and check if next ele is greater, if yes, will add in result, at max
//we'll traverse twice over input
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        Stack<Integer> st = new Stack<>();
        int n = nums.length;
        int[] result = new int[n];
        for (int i = 0; i < 2 * n; i++) {

            while (!st.isEmpty() && nums[i % n] > nums[st.peek()]) {
                result[st.peek()] = nums[i % n];
                st.pop();
            }
            if (i < n) {
                st.push(i);
                result[i % n] = -1;
            }

        }
        return result;
    }
}

File: NextGreaterEle.py
# Time Complexity : O(N)
# Space Complexity : O(N)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        
        st = []
        res = [-1] * len(nums)
        maxSoFar = float("-inf")
        
        for i in range(len(nums)):
            maxSoFar = max(maxSoFar, nums[i])
            while st and nums[i] > nums[st[-1]]:
                j = st.pop()
                res[j] = nums[i]
            st.append(i)
        
        if nums[st[-1]] < maxSoFar: # if top of stack is max no point in trying to resolve.
            for i in range(len(nums)):
                while st and nums[i] > nums[st[-1]]:
                    j = st.pop()
                    res[j] = nums[i]
                if nums[st[-1]] == maxSoFar: # if top of stack is max no point in trying to resolve. break out.
                    break
        
        return res